File: packages/ui-flutter/test/blocks/editor_toolbar_block_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:kinetix_ui/kinetix_ui.dart';

// The "Editor toolbar" block — see testimonial_block_test.dart for how block fixtures work.
//
// Two kinds of choice, two controls: the mode is one-of-two and swaps the surface, so it is a segmented
// control; the marks combine, so they are a toggle group. The marks disable in Preview, because offering a
// formatting action when there is nothing to format is offering an action that cannot happen.
//
// kx-block:start
class EditorFormat {
  const EditorFormat(this.id, this.name, this.glyph, this.shortcut);
  final String id;
  final String name;
  final String glyph;
  final String shortcut;
}

const List<EditorFormat> kEditorFormats = <EditorFormat>[
  EditorFormat('bold', 'Bold', 'B', 'B'),
  EditorFormat('italic', 'Italic', 'I', 'I'),
  EditorFormat('code', 'Code', '</>', 'E'),
];

class EditorToolbarBlock extends StatefulWidget {
  const EditorToolbarBlock({super.key});

  @override
  State<EditorToolbarBlock> createState() => _EditorToolbarBlockState();
}

class _EditorToolbarBlockState extends State<EditorToolbarBlock> {
  String _mode = 'write';
  final Set<String> _marks = <String>{'bold'};

  @override
  Widget build(BuildContext context) {
    final bool editing = _mode == 'write';

    return Padding(
      padding: const EdgeInsets.all(8),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: <Widget>[
              // Flexible, not a bare child: the segmented control lays its two items out with flex, and a
              // Row hands its non-flex children an unbounded width. The combination throws, and nothing
              // below it is ever laid out — including the toolbar's semantics.
              Flexible(
                child: KinetixSegmentedControl(
                  children: <Widget>[
                    KinetixSegmentedControlItem(
                      'Write',
                      selected: editing,
                      onTap: () => setState(() => _mode = 'write'),
                    ),
                    KinetixSegmentedControlItem(
                      'Preview',
                      selected: !editing,
                      onTap: () => setState(() => _mode = 'preview'),
                    ),
                  ],
                ),
              ),
              const SizedBox(width: 12),
              KinetixToggleGroup(
                children: kEditorFormats.map((EditorFormat format) {
                  final bool on = _marks.contains(format.id);
                  // The name goes INSIDE the control. KinetixToggle already declares button + toggled +
                  // enabled on its own container node; wrapping it in an excluding Semantics threw all of
                  // that away and left a label with no role. Labelling the child and hiding the glyph lets
                  // the component's own node carry name, role and state together — and `onChanged: null`
                  // is what makes Preview announce as disabled rather than merely look it.
                  return KinetixToggleGroupItem(
                    pressed: on,
                    onChanged: editing
                        ? (bool value) => setState(() => value ? _marks.add(format.id) : _marks.remove(format.id))
                        : null,
                    child: Semantics(
                      label: format.name,
                      excludeSemantics: true,
                      child: Text(format.glyph),
                    ),
                  );
                }).toList(),
              ),
            ],
          ),
          const SizedBox(height: 8),
          Row(
            children: kEditorFormats.map((EditorFormat format) {
              return Padding(
                padding: const EdgeInsets.only(right: 16),
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: <Widget>[
                    KinetixFieldDescription(format.name),
                    const SizedBox(width: 6),
                    const KinetixKbd('⌘'),
                    const SizedBox(width: 2),
                    KinetixKbd(format.shortcut),
                  ],
                ),
              );
            }).toList(),
          ),
        ],
      ),
    );
  }
}
// kx-block:end

void main() {
  testWidgets('formatting is named, and switches off with the editor', (WidgetTester tester) async {
    final SemanticsHandle handle = tester.ensureSemantics();
    await tester.pumpWidget(
      MaterialApp(
        home: KinetixTheme(
          brightness: Brightness.light,
          child: const Scaffold(body: SingleChildScrollView(child: EditorToolbarBlock())),
        ),
      ),
    );

    // Name, role and pressed state on one node. `.first` because the label is declared on the child and
    // merged up: both that Semantics and the toggle's own container resolve to the same node.
    expect(
      tester.getSemantics(find.bySemanticsLabel('Bold').first),
      isSemantics(label: 'Bold', isButton: true, isToggled: true, hasEnabledState: true, isEnabled: true),
    );

    await tester.ensureVisible(find.text('Preview'));
    await tester.pumpAndSettle();
    await tester.tap(find.text('Preview'));
    await tester.pump();

    // Still named and still a button — just not operable. A control that only greys out says nothing.
    expect(
      tester.getSemantics(find.bySemanticsLabel('Bold').first),
      isSemantics(label: 'Bold', isButton: true, hasEnabledState: true, isEnabled: false),
    );

    handle.dispose();
  });
}

File: packages/ui-compose/ui/src/test/kotlin/com/kinetixui/ui/blocks/EditorToolbarBlockTest.kt
package com.kinetixui.ui.blocks

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.test.assertIsNotEnabled
import androidx.compose.ui.test.assertIsOn
import androidx.compose.ui.test.junit4.v2.createComposeRule
import androidx.compose.ui.test.onNodeWithContentDescription
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import androidx.compose.ui.unit.dp
import com.kinetixui.ui.KinetixKbd
import com.kinetixui.ui.KinetixSegmentedControl
import com.kinetixui.ui.KinetixSegmentedControlItem
import com.kinetixui.ui.KinetixTheme
import com.kinetixui.ui.KinetixToggleGroup
import com.kinetixui.ui.KinetixToggleGroupItem
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

// The "Editor toolbar" block — see TestimonialBlockTest.kt for how block fixtures work.
//
// Two kinds of choice, two controls: the mode is one-of-two and swaps the surface, so it is a segmented
// control; the marks combine, so they are a toggle group. The marks disable in Preview, because offering a
// formatting action when there is nothing to format is offering an action that cannot happen.
//
// kx-block:start
private data class EditorFormat(val id: String, val name: String, val glyph: String, val shortcut: String)

private val EDITOR_FORMATS = listOf(
    EditorFormat("bold", "Bold", "B", "B"),
    EditorFormat("italic", "Italic", "I", "I"),
    EditorFormat("code", "Code", "</>", "E"),
)

@Composable
fun EditorToolbarBlock() {
    var mode by remember { mutableStateOf("write") }
    val marks = remember { mutableStateListOf("bold") }

    Column(
        modifier = Modifier.width(400.dp).padding(8.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            KinetixSegmentedControl {
                KinetixSegmentedControlItem("Write", selected = mode == "write", onClick = { mode = "write" })
                KinetixSegmentedControlItem("Preview", selected = mode == "preview", onClick = { mode = "preview" })
            }
            KinetixToggleGroup {
                EDITOR_FORMATS.forEach { format ->
                    KinetixToggleGroupItem(
                        pressed = format.id in marks,
                        onPressedChange = { on -> if (on) marks.add(format.id) else marks.remove(format.id) },
                        // The name is the announcement; the glyph is decoration. Without this TalkBack reads
                        // the letter "B", which is not the name of anything.
                        modifier = Modifier.semantics { contentDescription = format.name },
                        enabled = mode != "preview",
                    ) { Text(format.glyph) }
                }
            }
        }

        Row(horizontalArrangement = Arrangement.spacedBy(16.dp)) {
            EDITOR_FORMATS.forEach { format ->
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    Text(format.name)
                    KinetixKbd("⌘")
                    KinetixKbd(format.shortcut)
                }
            }
        }
    }
}
// kx-block:end

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class EditorToolbarBlockTest {
    @get:Rule
    val rule = createComposeRule()

    @Test
    fun formatting_is_named_and_switches_off_with_the_editor() {
        rule.setContent { KinetixTheme(darkTheme = false) { EditorToolbarBlock() } }
        rule.onNodeWithContentDescription("Bold").assertIsOn()

        rule.onNodeWithText("Preview").performClick()
        rule.onNodeWithContentDescription("Bold").assertIsNotEnabled()
    }
}
